src/net/protocol_agnostic/stream.rs

```diff
@@ -133,78 +136,101 @@ impl PaStream {
+                .and_then(move |(udp_socket, addr)| {
+                    trace!("udp rendezvous connect succeeded.");
+                    let (utp_socket, utp_listener) = {
+                        UtpSocket::from_socket(udp_socket, &handle).map_err(
+                            UtpRendezvousConnectError::IntoUtpSocket,
+                        )?
+                    };
+                    trace!("returning utp socket.");
+                    Ok((utp_socket, utp_listener, addr))
                 })
         };
 
+
         let ret = {
             pump_channels
-                .select2(connect)
-                .map_err(|either| match either {
-                    Either::A((err, _)) |
-                    Either::B((err, _)) => err,
-                })
-                .and_then(|either| match either {
-                    Either::A(((), connect)) => connect.into_boxed(),
-                    Either::B((stream, _)) => future::ok(stream).into_boxed(),
-                })
+            .while_driving(tcp_connect)
+            .while_driving(udp_connect)
+            .map_err(|((e, _tcp_connect), _udp_connect)| e)
+            .and_then(move |((their_pk, tcp_connect), udp_connect)| {
+                if our_pk > their_pk {
+                    let utp_connect = {
+                        udp_connect
+                        .and_then(|(utp_socket, _utp_listener, addr)| {
+                            utp_socket
+                                .connect(&addr)
+                                //.map_err(UtpRendezvousConnectError::UtpConnect)
+                                //.map(PaStream::Utp)
+                                .map_err(UtpRendezvousConnectError::UtpConnect)
+                                .map(PaStream::Utp)
+                        })
+                    };
+                    let connect = {
+                        tcp_connect
+                        .first_ok2(utp_connect)
+                        .map_err(|(tcp_err, utp_err)| {
+                            PaRendezvousConnectError::AllProtocolsFailed {
+                                tcp: Box::new(tcp_err),
+                                utp: Box::new(utp_err),
+                            }
+                        })
+                    };
+                    connect
+                    .and_then(|stream| {
+                        tokio_io::io::write_all(stream, b"CHOOSE")
+                        .map_err(PaRendezvousConnectError::SendChoose)
+                    })
+                    .map(|(stream, _buf)| stream)
+                    .into_boxed()
+                } else {
+                    let tcp_connect = tcp_connect.map(take_chosen);
+                    let utp_connect = {
+                        udp_connect
+                        .and_then(|(_utp_socket, utp_listener, addr)| {
+                            utp_listener
+                            .incoming()
+                            .filter(move |stream| stream.peer_addr() == addr)
```

Comment:
Does this filter connections to self? If so, how is this possible?

Reply:
No, it just makes sure we're accepting the connection from the address we hole-punched to.

Reply:
oh, good :)